src/passes/GlobalTypeOptimization.cpp

```diff
@@ -192,48 +198,131 @@ struct GlobalTypeOptimization : public Pass {
         vec[i] = true;
       }
 
-      // Process removability. We check separately for the ability to
-      // remove in a general way based on sub+super-propagated info (that is,
-      // fields that are not used in sub- or super-types, and so we can
-      // definitely remove them from all the relevant types) and also in the
-      // specific way that only works for removing at the end, which as
-      // mentioned above only looks at super-types.
+      // Process removability.
       std::set<Index> removableIndexes;
       for (Index i = 0; i < fields.size(); i++) {
-        if (!subSuper[i].hasRead) {
+        // If there is no read whatsoever, in either subs or supers, then we can
+        // remove the field. That is so even if there are writes (it would be a
+        // pointless "write-only field").
+        auto hasNoReadsAnywhere = !subSuper[i].hasRead;
+
+        // Check for reads or writes in ourselves and our supers. If there are
+        // none, then operations only happen in our strict subtypes, and those
+        // subtypes can define the field there, and we don't need it here.
+        auto hasNoReadsOrWritesInSupers = !sub[i].hasRead && !sub[i].hasWrite;
```

Comment:
It's a little confusing that to look at what the supertypes do, we look at `sub`. Are there better names we could give things here?

Reply:
Yeah, maybe that isn't clearest. The idea is that `sub` has data propagated **to** subtypes, so if there is nothing there, nothing from our **super**types was propagated to us. I renamed to `dataFromSupers`.